src/engine.ts

```diff
@@ -0,0 +1,207 @@
+const R = require('ramda');
+import uuid from '@root/uuid';
+import { Element, ElementKind } from '@root/elements';
+
+const PRIORITIES = {
+  [ElementKind[ElementKind.heading1]]: 4,
+  [ElementKind[ElementKind.heading2]]: 4,
+  [ElementKind[ElementKind.heading4]]: 4,
+  [ElementKind[ElementKind.heading4]]: 4,
+  [ElementKind[ElementKind.heading5]]: 4,
+  [ElementKind[ElementKind.heading6]]: 4,
+  [ElementKind[ElementKind.paragraph]]: 3,
+  [ElementKind[ElementKind.preformatted]]: 5,
+  [ElementKind[ElementKind.strong]]: 6,
+  [ElementKind[ElementKind.em]]: 6,
+  [ElementKind[ElementKind["list-item"]]]: 1,
+  [ElementKind[ElementKind["o-list-item"]]]: 1,
+  [ElementKind[ElementKind["group-list-item"]]]: 1,
+  [ElementKind[ElementKind["group-o-list-item"]]]: 1,
+  [ElementKind[ElementKind.image]]: 1,
+  [ElementKind[ElementKind.embed]]: 1,
+  [ElementKind[ElementKind.hyperlink]]: 3,
+  [ElementKind[ElementKind.label]]: 4,
+  [ElementKind[ElementKind.span]]: 7,
+};
+
+export interface Tree {
+  key: string;
+  children: Node[];
+}
+
+export interface Node {
+  key: string;
+  start: number;
+  end: number;
+  tag: string;
+  text?: string;
+  children: Node[];
+}
+
+class TextNode implements Node {
+  key: string;
+  start: number;
+  end: number;
+  tag: string;
+  text?: string;
+  children: Node[];
+
+  constructor(start: number, end: number, text: string) {
+    this.key = uuid();
+    this.start = start;
+    this.end = end;
+    this.tag = ElementKind[ElementKind.span];
+    this.text = text;
+    this.children = [];
+  }
+}
+
+export interface RichTextSpan {
+  start: number;
+  end: number;
+  type: string;
+  text: string;
+}
+
+export interface RichTextBlock {
+  text: string;
+  spans: RichTextSpan[];
+}
+
+export interface Group {
+  elected: Node;
+  others: Node[];
+}
+
+interface SlicedNode {
+  inner: Node;
+  outer?: Node;
+}
+
+function sliceNode(text: string, elected: Node, node: Node): SlicedNode {
+  if (node.start < elected.start) {
```

Comment:
Missing case? node.start < node.end && node.end > elected.end
should have two outer nodes in this case

Reply:
I don't think this case should happen because before that we grouped.